pc_hal/conioHistory.cpp

```diff
@@ -0,0 +1,280 @@
+/**
+ * @file conioHistory.cpp
+ * @author Ethan Huang (e8huang@ucsd.edu)
+ * @brief Command Line History Implementation
+ * @version 0.1
+ * @date 2025-03-06
+ *
+ * @copyright Copyright (c) 2025
+ *
+ */
+
+#include "conioHistory.hpp"
+
+#include "product.hpp"
+
+#if SF_PLATFORM == SF_PLATFORM_GLIBC
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+conioHistory::conioHistory() 
+{
+}
+
+conioHistory &conioHistory::getInstance(void)
+{
+    static conioHistory instance;
+    return instance;
+}
+
+conioHistory::CONIO_history_line_::CONIO_history_line_(std::size_t _offset) : offset(_offset), len(0), frag_seq(0), display(false), more_frag(false)
+{
+}
+
+conioHistory::CONIO_history_line_::CONIO_history_line_(std::size_t _offset,
+                                                       std::size_t _len,
+                                                       std::uint32_t _frag_seq,
+                                                       bool _display,
+                                                       bool _more_frag)
+    : offset(_offset), len(_len), frag_seq(_frag_seq), display(_display), more_frag(_more_frag)
+    {
+    }
+
+void conioHistory::init_file_mapping(void)
+{
+    const char *env_path = std::getenv("E4E_SF_CONIO_HISTORY");
+    env_path = env_path ? env_path : "/code/build/history.log";
+
+    fd = open(env_path, O_RDWR | O_CREAT | O_TRUNC, 0666);
+    if (fd < 0)
+    {
+        exit(1);
+    }
+
+    // Set initial file size
+    ftruncate(fd, INITIAL_FILE_SIZE);
+
+    // Map the file to memory
+    mapped_memory = (char *)mmap(nullptr, file_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
```

Comment:
It is worth considering creating a macro here so that mmaping of memory in two locations (here and in resize) cannot possibly come out of sync